Turbopack: Dynamic imports with complex patterns fail when styled-jsx wildcard exports are configured

### Link to the code that reproduces this issue

https://github.com/alexanderbnelson/dynamic-import-turbo

### To Reproduce

1. Clone the reproduction repository: `git clone https://github.com/alexanderbnelson/dynamic-import-turbo`
2. Install dependencies: `pnpm install`
3. Start the development server with Turbopack: `pnpm run dev` (runs `next dev --turbo`)
4. Navigate to `http://localhost:3000` in your browser
5. The error will appear in the terminal and browser

### Current vs. Expected behavior

**Current behavior:**
When using Turbopack with dynamic imports that have complex path patterns (double dynamic segments) and styled-jsx optimization configured, the following error occurs:

```
Module not found: Can't resolve '@/styles/themes/theme-' <dynamic> '/theme-' <dynamic>
  32 |   const themeName = site?.themeName ? toKebabCase(site?.themeName) : "one";
  33 |
> 34 |   const selectedTheme = await import(`@/styles/themes/theme-${themeName}/theme-${themeName}`);
     |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  35 |
  36 |   console.log(themeName);
  37 |   console.log(selectedTheme);

complex patterns into wildcard exports fields are not implemented yet: '@/styles/themes/theme-' <dynamic> '/theme-' <dynamic> into 'styled-jsx/*'

Debug info:
- Execution of <ModuleAssetContext as AssetContext>::resolve_asset failed
- Execution of resolve failed
- Execution of resolve_internal failed
- complex patterns into wildcard exports fields are not implemented yet: '@/styles/themes/theme-' <dynamic> '/theme-' <dynamic> into 'styled-jsx/*'
Error while looking up import map: complex patterns into wildcard exports fields are not implemented yet: '@/styles/themes/theme-' <dynamic> '/theme-' <dynamic> into 'styled-jsx/*'
```


https://nextjs.org/docs/messages/module-not-found`

**Expected behavior:** 
The dynamic import should resolve correctly and load the appropriate theme module, similar to how it works without Turbopack or without the styled-jsx configuration.

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 24.6.0: Mon Jul 14 11:30:29 PDT 2025; root:xnu-11417.140.69~1/RELEASE_ARM64_T6000
  Available memory (MB): 32768
  Available CPU cores: 10
Binaries:
  Node: 22.16.0
  npm: 11.4.2
  Yarn: 1.22.19
  pnpm: 10.16.0
Relevant Packages:
  next: 15.6.0-canary.6 // Latest available version is detected (15.6.0-canary.6).
  eslint-config-next: 15.5.3
  react: 19.1.1
  react-dom: 19.1.1
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Turbopack

### Which stage(s) are affected? (Select all that apply)

next dev (local)

### Additional context

Effects build also

This is my best attempt at a reproduction. My subject repo where this is occurring is a highly complex multi-tenant app. The error in this reproduction goes away when `optimizePackageImports: ['styled-jsx']` and `transpilePackages: ['styled-jsx']` are removed. But, in my subject repo, I get the very same error even without these config params. 

`complex patterns into wildcard exports fields are not implemented yet: '@/styles/themes/theme-' <dynamic> '/theme-' <dynamic> into 'styled-jsx/*'`

My subject project is calling for `styled-jsx` to be used but I'm not entirely sure where. The only place I see it referenced is in my lock file as a peer dependancy for `@babel/core`

Clearly maintainers are aware of this given the error message: `complex patterns into wildcard exports fields are not implemented yet`. I guess I'm wondering, is this avoidable? Will such a pattern be implemented any time soon?

<sub>[PACK-5451](https://linear.app/vercel/issue/PACK-5451/turbopack-dynamic-imports-with-complex-patterns-fail-when-styled-jsx)</sub>

--- a/turbopack/crates/turbopack-core/src/resolve/alias_map.rs
+++ b/turbopack/crates/turbopack-core/src/resolve/alias_map.rs
@@ -562,6 +562,10 @@ where
                         }
                     }
                     AliasKey::Wildcard { suffix } => {
+                        // This is quite messy as we'd have to match against the FS to do this
+                        // properly. For now, try to support as many cases as possible (and as are
+                        // actually used).
+
                         let is_match = if let Some(request) = self.request.as_constant_string() {
                             // The request is a constant string, so the PatriciaMap lookup already
                             // ensured that the prefix is matching the request.
@@ -587,9 +591,16 @@ where
                             ] = req.as_slice()
                         {
                             req_prefix.starts_with(&**prefix) && req_suffix.ends_with(&**suffix)
+                        } else if !self.request.could_match(prefix) {
+                            // There's no way it could match if the prefix can't match.
+                            false
+                        } else if suffix.is_empty() {
+                            // Prefix matches the request, and suffix is empty.
+                            true
                         } else {
+                            // It may or may not match, throw an error.
                             return Some(Err(anyhow::anyhow!(
-                                "complex patterns into wildcard exports fields are not \
+                                "Complex patterns into wildcard exports fields are not \
                                  implemented yet: {} into '{}*{}'",
                                 self.request.describe_as_string(),
                                 prefix,

--- a/turbopack/crates/turbopack-tests/tests/execution/turbopack/resolving/tsconfig-paths-dynamic/input/index.js
+++ b/turbopack/crates/turbopack-tests/tests/execution/turbopack/resolving/tsconfig-paths-dynamic/input/index.js
@@ -1,4 +1,4 @@
-import { loadSub, /* loadSubNested, */ loadSubFallback } from './src/foo'
+import { loadSub, loadSubNested, loadSubFallback } from './src/foo'
 
 it('should support dynamic requests with tsconfig.paths', () => {
   expect(loadSub('file2.js').default).toBe('file2')
@@ -8,9 +8,9 @@ it('should support dynamic requests with tsconfig.paths and without extension', 
   expect(loadSub('file2').default).toBe('file2')
 })
 
-// it('should support dynamic requests with tsconfig.paths and multiple dynamic parts', () => {
-//   expect(loadSubNested('file2').default).toBe('file2')
-// })
+it('should support dynamic requests with tsconfig.paths and multiple dynamic parts', () => {
+  expect(loadSubNested('file2').default).toBe('file2')
+})
 
 it('should support dynamic requests with tsconfig.paths fallbacks', () => {
   expect(loadSubFallback('file1').default).toBe('file1')

--- a/turbopack/crates/turbopack-tests/tests/execution/turbopack/resolving/tsconfig-paths-dynamic/input/src/foo.ts
+++ b/turbopack/crates/turbopack-tests/tests/execution/turbopack/resolving/tsconfig-paths-dynamic/input/src/foo.ts
@@ -2,10 +2,9 @@ export function loadSub(v: string) {
   return require(`@/sub/${v}`)
 }
 
-// TODO not supported
-// export function loadSubNested(v: string) {
-//   return require(`@/sub-nested/${v}/${v}.js`)
-// }
+export function loadSubNested(v: string) {
+  return require(`@/sub-nested/${v}/${v}.js`)
+}
 
 export function loadSubFallback(v: string) {
   return require(`@sub/${v}`)

